examples/mnist_denoising_autoencoder.py

```diff
@@ -0,0 +1,120 @@
+'''Trains a denoising autoenconder on MNIST dataset.
+
+Gets val_loss=0.02 after 1 epoch.
+Modular Encoder, Decoder and Autoencoder models.
+'''
+
+import keras
+from keras.layers import Activation, Dense, Input, BatchNormalization
+from keras.layers import Conv2D, MaxPooling2D, Flatten
+from keras.layers import Reshape, Conv2DTranspose, UpSampling2D
+from keras.models import Model
+from keras.datasets import mnist
+import numpy as np
+import matplotlib.pyplot as plt
+
+# MNIST dataset
+(x_train, _), (x_test, _) = mnist.load_data()
+
+image_size = x_train.shape[1]
+x_train = np.reshape(x_train, [-1, image_size, image_size, 1])
+x_test = np.reshape(x_test, [-1, image_size, image_size, 1])
+x_train = x_train.astype('float32') / 255
+x_test = x_test.astype('float32') / 255
+
+# Generate corrupted MNIST images by adding noise with normal dist
+# centered at 0.5 and std=0.5
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_train.shape)
+x_train_noisy = x_train + noise
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_test.shape)
+x_test_noisy = x_test + noise
+
+x_train_noisy = np.clip(x_train_noisy, 0., 1.)
+x_test_noisy = np.clip(x_test_noisy, 0., 1.)
+
+# Network parameters
+input_shape = (image_size, image_size, 1)
+batch_size = 128
+kernel_size = 3
+pool_size = 2
+filters = 16
+latent_dim = 16
+
+# Build the Autoencoder model
+# First build the Encoder model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of BN-ReLU-Conv2D-MaxPooling blocks
+for i in range(2):
+    x = BatchNormalization()(x)
+    x = Activation('relu')(x)
+    filters = filters * 2
+    x = Conv2D(filters=filters,
+               kernel_size=kernel_size,
+               padding='same')(x)
+    x = MaxPooling2D()(x)
+
+# Shape info needed to build decoder model
+shape = x.shape.as_list()
+
+# Generate a 16-dim latent vector
+x = Flatten()(x)
+latent = Dense(latent_dim, name='latent_vector')(x)
+
+# Instantiate Encoder model
+encoder = Model(inputs, latent, name='encoder')
+encoder.summary()
+
+# Build the Decoder model
+latent_inputs = Input(shape=(latent_dim,), name='decoder_input')
+x = Dense(shape[1] * shape[2] * shape[3])(latent_inputs)
+x = Reshape((shape[1], shape[2], shape[3]))(x)
+
+# Stack of BN-ReLU-Transposed Conv2D-UpSampling blocks
+for i in range(2):
+    x = BatchNormalization()(x)
+    x = Activation('relu')(x)
+    x = Conv2DTranspose(filters=filters,
+                        kernel_size=kernel_size,
+                        padding='same')(x)
+    x = UpSampling2D()(x)
```

Comment:
You don't need this, `Conv2DTranspose` with strides should work better

Reply:
Thanks. I included comments on the advantage/disadvantage of UpSampling2D.